{"id": "39699", "title": "Bug 39699simple deploy failed in cluster (distributable flag)", "description": " Bug 39699simple deploy failed in cluster (distributable flag) Hi,i encounter a strange problem with my tomcat cluster and webapps deployment.server.xml<Server port=\"8005\" shutdown=\"SHUTDOWN\"<Listener className=\"org.apache.catalina.core.AprLifecycleListener\" /<Listener className=\"org.apache.catalina.mbeans.ServerLifecycleListener\" /<ListenerclassName=\"org.apache.catalina.mbeans.GlobalResourcesLifecycleListener\" /<ListenerclassName=\"org.apache.catalina.storeconfig.StoreConfigLifecycleListener\"/<! Global JNDI resources <GlobalNamingResources<Environment name=\"simpleValue\" type=\"java.lang.Integer\" value=\"30\"/<Resource name=\"UserDatabase\" auth=\"Container\"type=\"org.apache.catalina.UserDatabase\"description=\"User database that can be updated and saved\"factory=\"org.apache.catalina.users.MemoryUserDatabaseFactory\"pathname=\"conf/tomcatusers.xml\" /</GlobalNamingResources<! Define the Tomcat StandAlone Service <Service name=\"Catalina\"<! Define a nonSSL HTTP/1.1 Connector on port 8080 <Connector address=\"poleninteg\" port=\"8080\" maxHttpHeaderSize=\"8192\"maxThreads=\"300\" minSpareThreads=\"25\" maxSpareThreads=\"100\"enableLookups=\"false\" acceptCount=\"100\"connectionTimeout=\"20000\" disableUploadTimeout=\"true\" /<! Define an AJP 1.3 Connector on port 8009 <Connector port=\"8009\"address=\"poleninteg.sofinco.fr\"minProcessors=\"20\" maxProcessors=\"0\" maxThreads=\"350\"minSpareThreads=\"50\" maxSpareThreads=\"100\"enableLookups=\"false\" redirectPort=\"8080\" protocol=\"AJP/1.3\"disableUploadTimeout=\"true\" /<Engine name=\"Catalina\" defaultHost=\"localhost\" jvmRoute=\"uxdev103\" <Realm className=\"org.apache.catalina.realm.UserDatabaseRealm\"resourceName=\"UserDatabase\"/<Host name=\"localhost\" appBase=\"webapps\" unpackWARs=\"true\" autoDeploy=\"true\"xmlValidation=\"false\" xmlNamespaceAware=\"false\"<! Cluster Definition <Cluster className=\"org.apache.catalina.cluster.tcp.SimpleTcpCluster\"doClusterLog=\"true\" clusterLogName=\"clusterlog\"expireSessionsOnShutdown=\"false\"useDirtyFlag=\"true\"notifyListenersOnReplication=\"true\"<Membership className=\"org.apache.catalina.cluster.mcast.McastService\"mcastAddr=\"228.0.0.4\"mcastBindAddress=\"127.0.0.1\"mcastClusterDomain=\"poleninteg\"mcastPort=\"45564\"mcastFrequency=\"1000\"mcastDropTime=\"30000\"/<Receiver className=\"org.apache.catalina.cluster.tcp.ReplicationListener\"tcpListenAddress=\"poleninteg\"tcpSelectorTimeout=\"100\"tcpThreadCount=\"6\"/<Sender className=\"org.apache.catalina.cluster.tcp.ReplicationTransmitter\"replicationMode=\"fastasyncqueue\"doTransmitterProcessingStats=\"true\"doProcessingStats=\"false\"doWaitAckStats=\"false\"waitForAck=\"false\"keepAliveMaxRequestCount=\"1\"/<Valve className=\"org.apache.catalina.cluster.tcp.ReplicationValve\"filter=\".\\.gif;.\\.js;.\\.css;.\\.png;.\\.jpeg;.\\.jpg;.\\.htm;.\\.html;.\\.txt;\"/<Valve className=\"org.apache.catalina.cluster.session.JvmRouteBinderValve\"enabled=\"true\" /<ClusterListenerclassName=\"org.apache.catalina.cluster.session.ClusterSessionListener\" /<ClusterListenerclassName=\"org.apache.catalina.cluster.session.JvmRouteSessionIDBinderListener\" /<Deployer className=\"org.apache.catalina.cluster.deploy.FarmWarDeployer\"tempDir=\"/home/tomcat/poleninteg/tmp/wartemp/\"deployDir=\"${catalina.base}/webapps\"watchDir=\"/home/tomcat/poleninteg/tmp/warlisten/\"watchEnabled=\"true\"/</Cluster</Host</Engine</Service</Serverwhen i am trying to deploy a 'dsitributable' (and only when this flag isenabled) i get the following errors :INFO ContainerBackgroundProcessor[StandardEngine[Catalina]]org.apache.catalina.startup.HostConfigDeploying web application archive polen.war INFO ContainerBackgroundProcessor[StandardEngine[Catalina]]org.apache.catalina.loader.WebappClassLoaderIllegal access: this web application instance has been stopped already.Could not load java.io.PrintStream.The eventual following stack trace is caused by an error thrown for debugging purposes as well as to attempt to terminate the thread whichcaused the illegal access, and has no functional impact. java.lang.IllegalStateExceptionatorg.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1238)atorg.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1198)at java.lang.ClassLoader.loadClassInternal(Unknown Source)at org.apache.log4j.helpers.LogLog.error(LogLog.java:142)atorg.apache.log4j.helpers.PatternParser$DatePatternConverter.convert(PatternParser.java:447)atorg.apache.log4j.helpers.PatternConverter.format(PatternConverter.java:64)at org.apache.log4j.PatternLayout.format(PatternLayout.java:503)at org.apache.log4j.WriterAppender.subAppend(WriterAppender.java:301)atorg.apache.log4j.RollingFileAppender.subAppend(RollingFileAppender.java:234)at org.apache.log4j.WriterAppender.append(WriterAppender.java:159)at org.apache.log4j.AppenderSkeleton.doAppend(AppenderSkeleton.java:230)atorg.apache.log4j.helpers.AppenderAttachableImpl.appendLoopOnAppenders(AppenderAttachableImpl.java:65)at org.apache.log4j.Category.callAppenders(Category.java:203)at org.apache.log4j.Category.forcedLog(Category.java:388)at org.apache.log4j.Category.log(Category.java:853)at org.apache.commons.logging.impl.Log4JLogger.info(Log4JLogger.java:133)atorg.apache.catalina.cluster.session.DeltaManager.start(DeltaManager.java:976)at org.apache.catalina.core.ContainerBase.setManager(ContainerBase.java:431)at org.apache.catalina.core.StandardContext.start(StandardContext.java:4125) at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:739)at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:524)at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:809)at org.apache.catalina.startup.HostConfig.deployWARs(HostConfig.java:698)at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:472)at org.apache.catalina.startup.HostConfig.check(HostConfig.java:1190)atorg.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:292)atorg.apache.catalina.util.LifecycleSupport.fireLifecycleEvent(LifecycleSupport.java:119)atorg.apache.catalina.core.ContainerBase.backgroundProcess(ContainerBase.java:1305)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1569)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1578)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.run(ContainerBase.java:1558)at java.lang.Thread.run(Unknown Source)INFO ContainerBackgroundProcessor[StandardEngine[Catalina]]org.apache.catalina.loader.WebappClassLoaderIllegal access: this web application instance has been stopped already.Could not load java.io.PrintStream.The eventual following stack trace is caused by an error thrown for debugging purposes as well as to attempt to terminate the thread whichcaused the illegal access, and has no functional impact. Java.lang.IllegalStateExceptionatorg.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1238)atorg.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1198)at java.lang.ClassLoader.loadClassInternal(Unknown Source)at org.apache.log4j.helpers.LogLog.error(LogLog.java:142)atorg.apache.log4j.helpers.PatternParser$DatePatternConverter.convert(PatternParser.java:447)atorg.apache.log4j.helpers.PatternConverter.format(PatternConverter.java:64)at org.apache.log4j.PatternLayout.format(PatternLayout.java:503)at org.apache.log4j.WriterAppender.subAppend(WriterAppender.java:301)atorg.apache.log4j.RollingFileAppender.subAppend(RollingFileAppender.java:234)at org.apache.log4j.WriterAppender.append(WriterAppender.java:159)at org.apache.log4j.AppenderSkeleton.doAppend(AppenderSkeleton.java:230)atorg.apache.log4j.helpers.AppenderAttachableImpl.appendLoopOnAppenders(AppenderAttachableImpl.java:65)at org.apache.log4j.Category.callAppenders(Category.java:203)at org.apache.log4j.Category.forcedLog(Category.java:388)at org.apache.log4j.Category.log(Category.java:853)at org.apache.commons.logging.impl.Log4JLogger.info(Log4JLogger.java:133)atorg.apache.catalina.cluster.session.DeltaManager.start(DeltaManager.java:976) at org.apache.catalina.core.ContainerBase.setManager(ContainerBase.java:431)at org.apache.catalina.core.StandardContext.start(StandardContext.java:4125)atorg.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:759)at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:739)at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:524)at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:809)at org.apache.catalina.startup.HostConfig.deployWARs(HostConfig.java:698)at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:472)at org.apache.catalina.startup.HostConfig.check(HostConfig.java:1190)atorg.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:292)atorg.apache.catalina.util.LifecycleSupport.fireLifecycleEvent(LifecycleSupport.java:119)atorg.apache.catalina.core.ContainerBase.backgroundProcess(ContainerBase.java:1305)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1569)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1578)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.run(ContainerBase.java:1558)at java.lang.Thread.run(Unknown Source)INFO ContainerBackgroundProcessor[StandardEngine[Catalina]]org.apache.catalina.loader.WebappClassLoaderIllegal access: this web application instance has been stopped already.Could not loadorg.apache.log4j.spi.VectorWriter.The eventual following stack trace is causedby an error thrown for debugging purposes as well as to attempt to terminate thethread which caused the illegal access, and has no functional impact. java.lang.IllegalStateExceptionatorg.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1238)atorg.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1198)at java.lang.ClassLoader.loadClassInternal(Unknown Source)at org.apache.log4j.spi.LoggingEvent.<init(LoggingEvent.java:154)at org.apache.log4j.Category.forcedLog(Category.java:388)at org.apache.log4j.Category.log(Category.java:853)at org.apache.commons.logging.impl.Log4JLogger.error(Log4JLogger.java:193)atorg.apache.catalina.cluster.session.DeltaManager.start(DeltaManager.java:990)at org.apache.catalina.core.ContainerBase.setManager(ContainerBase.java:431)at org.apache.catalina.core.StandardContext.start(StandardContext.java:4125)atorg.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:759)at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:739)at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:524)at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:809)at org.apache.catalina.startup.HostConfig.deployWARs(HostConfig.java:698)at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:472) at org.apache.catalina.startup.HostConfig.check(HostConfig.java:1190)atorg.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:292)atorg.apache.catalina.util.LifecycleSupport.fireLifecycleEvent(LifecycleSupport.java:119)atorg.apache.catalina.core.ContainerBase.backgroundProcess(ContainerBase.java:1305)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1569)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1578)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.run(ContainerBase.java:1558)at java.lang.Thread.run(Unknown Source)INFO ContainerBackgroundProcessor[StandardEngine[Catalina]]org.apache.catalina.loader.WebappClassLoaderIllegal access: this web application instance has been stopped already.Could not loadorg.apache.log4j.spi.VectorWriter.The eventual following stack trace is causedby an error thrown for debugging purposes as well as to attempt to terminate thethread which caused the illegal access, and has no functional impact. java.lang.IllegalStateExceptionatorg.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1238)atorg.apache.catalina.loader.WebappClassLoader.loadClass(WebappClassLoader.java:1198)at java.lang.ClassLoader.loadClassInternal(Unknown Source)at org.apache.log4j.spi.LoggingEvent.<init(LoggingEvent.java:154)at org.apache.log4j.Category.forcedLog(Category.java:388)at org.apache.log4j.Category.log(Category.java:853)at org.apache.commons.logging.impl.Log4JLogger.error(Log4JLogger.java:193)atorg.apache.catalina.cluster.session.DeltaManager.start(DeltaManager.java:990)at org.apache.catalina.core.ContainerBase.setManager(ContainerBase.java:431)at org.apache.catalina.core.StandardContext.start(StandardContext.java:4125)atorg.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:759)at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:739)at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:524)at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:809)at org.apache.catalina.startup.HostConfig.deployWARs(HostConfig.java:698)at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:472)at org.apache.catalina.startup.HostConfig.check(HostConfig.java:1190)atorg.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:292)atorg.apache.catalina.util.LifecycleSupport.fireLifecycleEvent(LifecycleSupport.java:119)atorg.apache.catalina.core.ContainerBase.backgroundProcess(ContainerBase.java:1305)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1569)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1578)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.run(ContainerBase.java:1558)at java.lang.Thread.run(Unknown Source)ERROR ContainerBackgroundProcessor[StandardEngine[Catalina]]org.apache.commons.modeler.RegistryNull component Catalina:type=JspMonitor,name=jsp,WebModule=//localhost/polen,J2EEApplication=none,J2EEServer=none ERROR ContainerBackgroundProcessor[StandardEngine[Catalina]]org.apache.catalina.startup.HostConfigError deploying web application archive polen.war java.lang.NoClassDefFoundError: org/apache/log4j/spi/VectorWriterat org.apache.log4j.spi.LoggingEvent.<init(LoggingEvent.java:154)at org.apache.log4j.Category.forcedLog(Category.java:388)at org.apache.log4j.Category.log(Category.java:853)at org.apache.commons.logging.impl.Log4JLogger.error(Log4JLogger.java:193)atorg.apache.catalina.cluster.session.DeltaManager.start(DeltaManager.java:990)at org.apache.catalina.core.ContainerBase.setManager(ContainerBase.java:431)at org.apache.catalina.core.StandardContext.start(StandardContext.java:4125)atorg.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:759)at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:739)at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:524)at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:809)at org.apache.catalina.startup.HostConfig.deployWARs(HostConfig.java:698)at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:472)at org.apache.catalina.startup.HostConfig.check(HostConfig.java:1190)atorg.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:292)atorg.apache.catalina.util.LifecycleSupport.fireLifecycleEvent(LifecycleSupport.java:119)atorg.apache.catalina.core.ContainerBase.backgroundProcess(ContainerBase.java:1305)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1569)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1578)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.run(ContainerBase.java:1558)at java.lang.Thread.run(Unknown Source)ERROR ContainerBackgroundProcessor[StandardEngine[Catalina]]org.apache.catalina.startup.HostConfigError deploying web application archive polen.war java.lang.NoClassDefFoundError: org/apache/log4j/spi/VectorWriterat org.apache.log4j.spi.LoggingEvent.<init(LoggingEvent.java:154)at org.apache.log4j.Category.forcedLog(Category.java:388)at org.apache.log4j.Category.log(Category.java:853)at org.apache.commons.logging.impl.Log4JLogger.error(Log4JLogger.java:193)atorg.apache.catalina.cluster.session.DeltaManager.start(DeltaManager.java:990)at org.apache.catalina.core.ContainerBase.setManager(ContainerBase.java:431)at org.apache.catalina.core.StandardContext.start(StandardContext.java:4125)at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:759)at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:739)at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:524)at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:809)at org.apache.catalina.startup.HostConfig.deployWARs(HostConfig.java:698)at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:472)at org.apache.catalina.startup.HostConfig.check(HostConfig.java:1190)atorg.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:292)atorg.apache.catalina.util.LifecycleSupport.fireLifecycleEvent(LifecycleSupport.java:119)atorg.apache.catalina.core.ContainerBase.backgroundProcess(ContainerBase.java:1305)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1569)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1578)atorg.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.run(ContainerBase.java:1558)at java.lang.Thread.run(Unknown Source)i have tried to deployed via web manager application or simple copy of my war inwebapps directory... No luck, if i put '<distributable/ flag in applicationweb.xml config file, i encounter this error ; i need to restart manually tomcatinstance... After this restart the application is correctly deployed without anyerror...Of course if i disable cluster in server.xml, no problem...", "OB": "", "EB": "", "SR": ""}